Fix type error when converting interpolated_linear_operator to double precision

CY-Zhang opened the pull request:
- Fix type error when converting interpolated_linear_operator to double precision, as discussed in https://github.com/cornellius-gp/gpytorch/issues/1958
- Added to() method in interpolated_linear_operator.py to overwrite the default to() method in _linear_operator. For each arg, the dtype will be converted only when the original dtype and target dtype are both int/float, to avoid converting integer index matrices to float.
- Added new unit test case under linear_operator_test_case.py to test whether after calling to(torch.float64), the linear operator is still functional and can be converted to numpy array. Raise TypeError if the test failed.

Commit:
fixed type error when calling to() method

Commit:
Added unit test case for dtype conversion to double.

Commit:
modified error message for unit test.

Balandat commented on linear_operator/operators/interpolated_linear_operator.py:
```diff
@@ -454,3 +455,28 @@ def zero_mean_mvn_samples(
+        new_kwargs = {}
+        for arg in self._args:
+            if hasattr(arg, "to"):
+                if arg.dtype.is_floating_point == dtype.is_floating_point:
```
It could be that `arg` has a `to` attribute but doesn't have a `dtype` attribute (e.g. if it is some other non-tensor or non-module object that just happens to have a `to` method). Not sure exactly how we'd handle this but we probably should? Maybe we want to catch and pass on exceptions here?

CY-Zhang replied:
That's a good point. Maybe I can add another condition to first check if the `arg` has a `dtype` attribute, before the `is_floating_point` check? If the `arg` does not have a `dtype` attribute, we directly jump to the `device` conversion.

Balandat commented on linear_operator/test/linear_operator_test_case.py:
```diff
@@ -1247,3 +1247,12 @@ def test_svd(self):
+        # test if the linear_op is still functional after convert to double
+        linear_op = self.create_linear_op()
+        try:
+            linear_op = linear_op.to(torch.float64)
+            linear_op.numpy()
```
Should we also test that the dtype was properly updated? 

Also, we probably want to overwrite this for the interpolated operator to verify that the index tensors were not moved to floating point type

CY-Zhang replied:
Make sense, I will include test to check if dtype is updatedm, and check if index tensors are still integers for the interpolated operator case.

Balandat reviewed (commented):
Thanks for putting this together!

Commit:
updated unit test and to() method

Balandat commented on linear_operator/test/linear_operator_test_case.py:
```diff
@@ -1247,3 +1247,14 @@ def test_svd(self):
+            linear_op = linear_op.to(torch.float64)
+            linear_op.numpy()
+        except RuntimeError:
+            raise TypeError(f"Could not convert {type(linear_op)} to double.")
```
This seems more like a `RuntimeError` than like a `TypeError`
```suggestion
            raise RuntimeError(f"Could not convert {type(linear_op)} to double.")
```

Balandat commented on test/operators/test_interpolated_linear_operator.py:
```diff
@@ -48,6 +48,23 @@ def evaluate_linear_op(self, linear_op):
+            linear_op = linear_op.to(torch.float64)
+            linear_op.numpy()
+        except RuntimeError:
+            raise TypeError(f"Could not convert {type(linear_op)} to double.")
```
```suggestion
            raise RuntimeError(f"Could not convert {type(linear_op)} to double.")
```

Balandat commented on test/operators/test_interpolated_linear_operator.py:
```diff
@@ -48,6 +48,23 @@ def evaluate_linear_op(self, linear_op):
+            linear_op.left_interp_indices.dtype == torch.float64
+            or linear_op.right_interp_indices.dtype == torch.float64
+        ):
+            raise TypeError(f"Index tensor in {type(linear_op)} converted to double.")
```
You probably want to use sth like this here to be consistent with the general unit test setup
```suggestion
        self.assertFalse(linear_op.left_interp_indices.dtype.is_floating_point)
        self.assertFalse(linear_op.right_interp_indices.dtype.is_floating_point)
```

Balandat commented on linear_operator/test/linear_operator_test_case.py:
```diff
@@ -1247,3 +1247,14 @@ def test_svd(self):
+        except RuntimeError:
+            raise TypeError(f"Could not convert {type(linear_op)} to double.")
+        if linear_op.dtype != torch.float64:
+            raise TypeError(f"Could not convert {type(linear_op)} to double.")
```
```suggestion
        self.assertEqual(linear_op.dtype, torch.float64)
```

CY-Zhang committed:
Apply suggestions from code review

Modified error types and revised error message to follow unittest framwork format.

Co-authored-by: Max Balandat <Balandat@users.noreply.github.com>

Balandat reviewed (commented):
Looks like there are some test failures; once those are fixed this is ready to go in.

CY-Zhang commented:
> Looks like there are some test failures; once those are fixed this is ready to go in.

Yeah, it seems besides the interpolated_linear_operator, the dtype conversion method will also cause some error when being applied to masked_linear_operator (update boolean mask matrices to double) and identity_linear_operator (dtype saved in kwargs and did not get updated). I will do similar fix as the interpolated_linear_operator on these two linear_operators.

Commit:
Fixed further issues during dtype conversion to double.

Balandat commented:
Thanks for tying up the loose ends here
